### DISCUSSION TITLE
Spring-Boot @Recurring and JobContext

### DISCUSSION THREAD

**Discussion by jeremylong** [jobrunr/jobrunr#409] [Q&A] [answered]
I am trying to figure out how a Recurring job can access the JobContext? I have a recurring job and would like to be able to access the JobContext so that I can add a `JobDashboardProgressBar`:

```
	@Recurring(id = "my-recurring-job", cron = "0 1 * * *")
	@Job(name = "sample-pipeline", retries = 2)
	public void executeePipeline() {
		JobContext ctx;
		var targets = determineTargets();
		var progressBar = ctx.progressBar(targets.size());
		// do stuff
	}
```

Any suggestions?

**Comment by rdehuyss** [comment, marked as answer]
That is use case I did not think of yet.

What I would do for now: 
- create a recurring job which enqueues a new job
- in the enqueued job, you have access to the job context.

This is how I typically do it by the way.  The recurring job checks whether there is work to do, and if so, it creates an actual job. For performance reasons the recurring job is instantly deleted using the custom delete policy of JobRunr Pro.

I'll also have a look to add the possibility of adding the job context to a recurring job.

**Comment by jeremylong** [comment]
In a review of the code, and I've not spent a ton of time looking at the jobrunr code, I've seen the following used to kick off jobs:
https://github.com/jobrunr/jobrunr/blob/dc73c63f382dbd4e41cc0629bdf318e5197c31a2/core/src/main/java/org/jobrunr/server/runner/AbstractBackgroundJobRunner.java#L53-L60

As it appears kicking off a job already looks for and injects the `JobContext` into the parameters - wouldn't just adjusting the validation in the `org.jobrunr.spring.annotations.Recurring` annotation processor to allow a parameter if it is `JobContext`?

Specifically: https://github.com/jobrunr/jobrunr/blob/d242a7f9e7073211d84b3f73a04eb6c149e9b90f/framework-support/jobrunr-spring-boot-starter/src/main/java/org/jobrunr/scheduling/RecurringJobPostProcessor.java#L59-L66

Could be updated to check for the `JobContext` as a single parameter:

```Java
@Override
public void doWith(Method method) throws IllegalArgumentException {
    if (!method.isAnnotationPresent(Recurring.class)) {
        return;
    }
    if (method.getParameterCount() > 0 && !(method.getParameterCount()==1 && JobConext.class.isAssignableFrom(method.getParameterTypes()[0]))) {
        throw new IllegalStateException("Methods annotated with " + Recurring.class.getName() + " can not have parameters.");
    }
```

Would this work? If so, I could create an issue and PR to resolve the issue (not sure how formal the project is with having all PRs linked to issues).

**Reply by rdehuyss** [reply]
I'm looking into it.

**Comment by rdehuyss** [comment]
That was the idea I also had in mind. 

Just a PR is fine. However, the same needs to be done for micronaut and quarkus.

**Comment by essentialprogramming** [comment]
I see the suggested change is already there: 

``
private boolean hasParametersOutsideOfJobContext(Method method) {
            if(method.getParameterCount() == 0) return false;
            else if(method.getParameterCount() > 1) return true;
            else return !method.getParameterTypes()[0].equals(JobContext.class);
        }
``

The comment in the annotation also suggests it should work: 

>  * <em>Note that methods annotated with the &commat;Recurring annotation may only have zero parameters or a single parameter of type org.jobrunr.jobs.context.JobContext.</em>


But is still does not seem to work. Should it work now ?

**Reply by rdehuyss** [reply]
Indeed, something looks off. I'm looking into it.

**Reply by rdehuyss** [reply]
See issue #468

**Reply by essentialprogramming** [reply]
Thank you! The update will be available in 5.1.3, I suppose. Is that correct ?

**Reply by rdehuyss** [reply]
Yes.

**Reply by rdehuyss** [reply]
Please star the repo if you can... it keeps me motivated ;-)
